gh19010: fix returns for Perl_grok_infnan

hvds opened the pull request:
Consistently honour what the docs have always promised:
  If an infinity or a not-a-number is recognized, C<*sp> will point to
  one byte past the end of the recognized string.  If the recognition fails,
  zero is returned, and C<*sp> will not move.

Additionally, restore Perl_grok_number_flags to allowing inf/nan with
trailing garbage only when called with PERL_SCAN_TRAILING; add notes
to the other two core callers to clarify that they always accept such
trailing garbage.

A small number of XS-APItest tests were modified to reflect the stricter
behaviour: "Infin" and "nanx" are now invalid without PERL_SCAN_TRAILING.

hvds committed:
gh19010: fix returns for Perl_grok_infnan

Consistently honour what the docs have always promised:
  If an infinity or a not-a-number is recognized, C<*sp> will point to
  one byte past the end of the recognized string.  If the recognition fails,
  zero is returned, and C<*sp> will not move.

Additionally, restore Perl_grok_number_flags to allowing inf/nan with
trailing garbage only when called with PERL_SCAN_TRAILING; add notes
to the other two core callers to clarify that they always accept such
trailing garbage.

A small number of XS-APItest tests were modified to reflect the stricter
behaviour: "Infin" and "nanx" are now invalid without PERL_SCAN_TRAILING.

hvds commented:
I'm not sure who is best placed to review this, but I'm asking @Perl/perl-steering-council for blessing as well since the work requires test changes.

khwilliamson commented:
@sisyphus what do you think?

rjbs commented:
(I am not competent to have a ready opinion here.)

hvds commented:
@t-a-k you've made some improvements to inf/nan handling in some areas recently, would you be able to take a look at the code here?

t-a-k commented:
It looks basically good to me, but there are some points to be worried about:
* Trailing spaces after inf/nan are normally ignored (and exempt from "isn't numeric" warnings), but spaces after "`1.#IND`" are not.  Is this intentional?
* `grok_infnan` would eventually have duplicated codes for `while (s < send && isSPACE(*s)) s++; *sp = s; return flags | (s < send ? IS_NUMBER_TRAILING : 0);`. I  suggest to consolidate them into one place, e.g. at the end of this function.

hvds committed:
gh19010: regularize handling of trailing whitespace on inf/nan

Always accept trailing whitespace when parsing Inf/NaN constructs, even
inside parens, without flagging as "trailing garbage".

Also fix off-by-one checking whether we reached the close-paren of
'nan(123)'.

hvds committed:
gh19010: simplify and expand inf/nan tests

We know how grok results should vary with and without PERL_SCAN_TRAILING
both for inputs with trailing garbage and those without, so abstract
that out and test both cases for each input.

hvds commented:
> It looks basically good to me, but there are some points to be worried about:
> 
> * Trailing spaces after inf/nan are normally ignored (and exempt from "isn't numeric" warnings), but spaces after "`1.#IND`" are not.  Is this intentional?
> 
> * `grok_infnan` would eventually have duplicated codes for `while (s < send && isSPACE(*s)) s++; *sp = s; return flags | (s < send ? IS_NUMBER_TRAILING : 0);`. I  suggest to consolidate them into one place, e.g. at the end of this function.

Thanks very much @t-a-k, I'd missed the `"1.#IND"` case, and looking more closely showed further cases to consider; that then also led me to a (minor) off-by-one error; and expanding the tests for this then showed up an additonal unrelated inconsistency in `Perl_grok_number_flags` (#19464).

I've pushed two additional commits, one to fix up the code and the second to abstract out the tests so that we check every case both with and without `PERL_SCAN_TRAILING`. In case it's useful I've also pushed a rebased branch [hv/gh19010-rebase](https://github.com/Perl/perl5/tree/hv/gh19010-rebase) with the commits squashed.

Please take another look; I'm interested to know also whether the _motivation_ of this PR makes sense to you, ie the assumption that what the docs promise is sane and useful, and that we should fix the code to match the docs.

t-a-k commented:
> Please take another look; I'm interested to know also whether the _motivation_ of this PR makes sense to you, ie the assumption that what the docs promise is sane and useful, and that we should fix the code to match the docs.

`grok_infnan` is publicly (in `perlapi`) documented to update `*sp` when it recognized Inf/NaN, so I think that failing to update `*sp` properly should be considered as a bug to be fixed; therefore I agree with the motivation of this PR.

t-a-k commented on numeric.c:
```diff
@@ -771,36 +771,29 @@ Perl_grok_infnan(pTHX_ const char** sp, const char* send)
-            if (s < send && *s) {
-                flags |= IS_NUMBER_TRAILING;
-            }
+            goto ok_check_space;
```
Are these `goto ok_check_space;`'s really needed?  Enclosing blocks seem to naturally exit to `ok_check_space:` even without them.

hvds replied:
I think they're valuable for consistency: the intent is that `goto ok_check_space` confirms we have met its prerequisites: "We parsed something valid, s points after it, flags describes it", hence also the `/* NOTREACHED */` to discourage simply falling out of the block. I'd expect any C compiler to generate the same code either way, this is just for the benefit of humans.

t-a-k replied:
Oh, that makes sense.
Then, how about using `NOT_REACHED` macro in addition to `/* NOTREACHED */`?

hvds replied:
> Oh, that makes sense. Then, how about using `NOT_REACHED` macro in addition to `/* NOTREACHED */`?

Ah yes, I had forgotten that we also had a macro for it. Now added, and also updated hv/gh19010-rebase.

hvds committed:
gh19010: add NOT_REACHED

hvds commented:
Thanks @khwilliamson, and thanks @t-a-k for helping to unblock this.
